### PR TITLE
[FLINK-2731][web-dashboard] Adds access to Job Manager stdout and log.

### PR DESCRIPTION
**Pull request by sachingoel0101**
This adds minimalist support for accessing Job manager logs and stdout files to the new web frontend.
Right now, to refresh the logs, user needs to explicitly press the refresh button, at which the whole log/stdout file is re-fetched and shown.

Later on, I plan to implement a better system which doesn't have to fetch the whole file again. That might take some time and not be complete in time for the next release.
@uce , can you take a look at this?
Attaching a few screenshots:
Logs:

log

Logs file download:

log_download

Stdout:

out

Stdout file download:

out_download

### CODE REVIEW

**Comment by uce** [comment]
Yes, I will look at it today. Thanks! Looks great from the screenshots.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java [hunk 1]
```diff
@@ -276,4 +294,18 @@ public int getServerPort() {
 	private static RuntimeMonitorHandler handler(RequestHandler handler) {
 		return new RuntimeMonitorHandler(handler);
 	}
+
+	private static String getFilePath(String pattern) {
```
Comment (fhueske) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: can you rename the function to `getLogFilePath()` or add `directory` parameter?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java [hunk 2]
```diff
@@ -123,9 +124,18 @@ public void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Except
 		final HttpRequest request = routed.request();
 		String requestPath = routed.path();
 
+		// TODO REMOVE BEFORE COMMITTING. THIS IS ONLY TO TEST THAT JOB MANAGER STDOUT IS ACCESSED
```
Comment (fhueske) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: Please remove

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java (hunk 1 above)
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: FYI I'm testing this locally and will change some minor stuff.

Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: There is a  list files version which takes a filter. You can directly use that one.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java [hunk 3]
```diff
@@ -160,8 +176,10 @@ else if (flinkRoot != null) {
 			// the overview - how many task managers, slots, free slots, ...
 			.GET("/overview", handler(new ClusterOverviewHandler(retriever, DEFAULT_REQUEST_TIMEOUT)))
 
-			// job manager configuration
+			// job manager configuration, log and stdout
 			.GET("/jobmanager/config", handler(new JobManagerConfigHandler(config)))
+			.GET("/jobmanager/logfile", new StaticFileServerHandler(logStdoutDir))
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: log is fine imo

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java [hunk 4]
```diff
@@ -139,7 +148,14 @@ else if (flinkRoot != null) {
 			throw new IllegalConfigurationException("The path to the static contents (" + 
 					webRootDir.getAbsolutePath() + ") is not a readable directory.");
 		}
-		
+
+		final File logStdoutDir = new File(flinkRoot, LOG_OUTPUT_DIRECTORY);
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: i like `logDir` is more consistent

Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: please use the same kind of checks as for the resources. its fails for my machine after building:
```
org.apache.flink.configuration.IllegalConfigurationException: The path to the log and stdout directory (/Users/ufuk/code/flink-2/flink-dist/target/flink-0.10-SNAPSHOT-bin/flink-0.10-SNAPSHOT/conf/../log) is not a readable directory.
	at org.apache.flink.runtime.webmonitor.WebRuntimeMonitor.<init>(WebRuntimeMonitor.java:156)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.apache.flink.runtime.jobmanager.JobManager$.startWebRuntimeMonitor(JobManager.scala:1772)
	at org.apache.flink.runtime.jobmanager.JobManager$.runJobManager(JobManager.scala:1322)
	at org.apache.flink.runtime.jobmanager.JobManager$.main(JobManager.scala:1194)
	at org.apache.flink.runtime.jobmanager.JobManager.main(JobManager.scala)
```

**Comment by fhueske** [comment]
Just found two minor issues, but I'm not a web guy and have no clue of coffee script and .jade files...

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java [hunk 5]
```diff
@@ -123,9 +124,18 @@ public void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Except
...
+		System.out.println(requestPath);
+
 		// make sure we request the "index.html" in case there is a directory request
 		if (requestPath.endsWith("/")) {
 			requestPath = requestPath + "index.html";
+		} else if (requestPath.equals("/jobmanager/logfile")) {
+			requestPath = "/" + WebRuntimeMonitor.LOG_FILE_PATH;
+		} else if (requestPath.equals("/jobmanager/stdout")) {
+			requestPath = "/" + WebRuntimeMonitor.STDOUT_FILE_PATH;
+		} else {
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: the else comment is confusing. please remove

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java [hunk 6]
```diff
@@ -177,7 +187,11 @@ public void channelRead0(ChannelHandlerContext ctx, Routed routed) throws Except
 
 		HttpResponse response = new DefaultHttpResponse(HTTP_1_1, OK);
 		setContentTypeHeader(response, file);
-		setDateAndCacheHeaders(response, file);
+		// TODO This is counter-intuitive. Will be changed later when a dynamic file server is added
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: please remove this special branch. what is confusing exactly?

File: flink-runtime-web/web-dashboard/app/styles/index.styl [hunk 7]
```diff
@@ -405,3 +405,7 @@ pre.exception
   .table
     td.td-long
       width: 30%
+
+a
```
Comment (uce) [inline flink-runtime-web/web-dashboard/app/styles/index.styl]: unrelated change. please revert

**Comment by uce** [comment]
I made some comments. I've initially said that I am going to address them myself, but I won't have the time today. Can you please address them? Thanks for the fast PR btw.

And how does this interact with YARN?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java (hunk 6 above)
Comment (sachingoel0101) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: This is to prevent the browser from caching the log or out files since they're highly dynamic. The refresh button will have no effect whatsoever otherwise.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java (hunk 5 above)
Comment (sachingoel0101) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: Will do.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java (hunk 4 above)
Comment (sachingoel0101) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java]: Yes. Moving around the log file path finding code will take care of this.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java (hunk 2 above)
Comment (sachingoel0101) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/files/StaticFileServerHandler.java]: will do

File: flink-runtime-web/web-dashboard/app/styles/index.styl (hunk 7 above)
Comment (sachingoel0101) [inline flink-runtime-web/web-dashboard/app/styles/index.styl]: This is kind of required to make the mouse change to pointer on hovering over the refresh button. But yes, will make it a little better.

**Comment by sachingoel0101** [comment]
@uce , @fhueske thanks for the review. 
I've addressed all your concerns and am testing it right now.
However, I'm not at all familiar with yarn and whether this will work with it or not. I'll see if I can setup yarn locally and test this out.

**Comment by sachingoel0101** [comment]
@uce, the new web dashboard doesn't work with yarn yet. The files are not shipped with the jar. I'm confused why it isn't being done so. The old frontend's files are included in the jar itself. 

Also, I may have missed something in configuring things but I don't think the old one works either.

**Comment by rmetzger** [comment]
I have a pending pull request to enable it on YARN.

**Comment by sachingoel0101** [comment]
@rmetzger , I looked at your changes to enable the dashboard on yarn. Looks good. :)

A few comments to make sure this PR can directly start working after yours:
I would like to rename the yarn log files to `jobmanager.log`, `jobmanager.out` and `jobmanager.err`. Can you do this with your commit?
Further, the web runtime resources and log file paths are entirely different things. Right now, in normal mode, I can assume that the log files are present under the directory `flink.base.dir.key/log`. However, that won't be the case in yarn. What would be the best way to pass the log directory path to Web runtime? Should it be given as an argument or as an environment variable? I'm more inclined towards an ev.

After you address my comments, I will rebase this on top of your PR and check if everything works. :)

**Comment by sachingoel0101** [comment]
@uce Since the web frontend isn't integrated yet, I have verified correct logs directory paths by examining container logs by deploying on yarn. This will now work with Yarn too.

**Comment by rmetzger** [comment]
@sachingoel0101: my pull request (https://github.com/apache/flink/pull/1222) is actually based on @uce's PR (https://github.com/apache/flink/pull/1202). I'm not sure when I'll find time to merge my PR.

**Comment by sachingoel0101** [comment]
IMO this can be merged as it works as expected in non-yarn mode. And the
number of changes to be made are minimal when you finally merge your PR.
I can remove the yarn specific changes in this and it should be mergeable
for now. What do you think?

**Comment by sachingoel0101** [comment]
@mxm I've rebased on your latest changes to enable dashboard on yarn. 
Also, there was a little redundant code in `StaticFileServerHandler` 
```java
if (File.separatorChar != '/') {
    requestPath = requestPath.replace('/', File.separatorChar);
}
```
which I think was meant to make sure things work on windows, but it doesn't actually serve that purpose. I've removed it in the second commit.

The log and out files are now accessible both with and without yarn. I have also enabled the logs related tests on both `YARNSessionFIFOITCase` and `WebFrontendITCase`.

**Comment by mxm** [comment]
Thanks a lot @sachingoel0101. I'll check out and test your changes now.

File: flink-tests/src/test/java/org/apache/flink/test/web/WebFrontendITCase.java [hunk 8]
```diff
@@ -110,21 +112,18 @@ public void getTaskmanagers() {
...
-//			Assert.fail(e.getMessage());
-//		}
-//	}
+	@Test
+	public void getLogAndStdoutFiles() {
+		try {
+			String logs = getFromHTTP("http://localhost:" + port + "/jobmanager/log");
+			Assert.assertTrue(logs.contains("job manager log"));
+			logs = getFromHTTP("http://localhost:" + port + "/jobmanager/stdout");
+			Assert.assertTrue(logs.contains("job manager out"));
```
Comment (mxm) [inline flink-tests/src/test/java/org/apache/flink/test/web/WebFrontendITCase.java]: How about writing something to the log or stdout and checking for the result here?

**Comment by mxm** [comment]
Looks good, @sachingoel0101. I think we can merge this if we adapt the test which checks for the log file access in the web frontend.

**Comment by sachingoel0101** [comment]
Hi @mxm I have fixed the `WebFrontendITCase`. Please have a look again.

**Comment by mxm** [comment]
Thanks. Looks great. Merging this.
